<p style="font-family: Arial; font-size:3em;color:black;"> Lab Exercise 6</p>

In [51]:
# you can download the original Kyphosis dataset from kaggle at https://www.kaggle.com/abbasit/kyphosis-dataset?select=kyphosis.csv

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [52]:
Kyphosis_df = pd.read_csv('kyphosis.csv')
print(Kyphosis_df.head(10))

  Kyphosis  Age  Number  Start
0   absent   71       3      5
1   absent  158       3     14
2  present  128       4      5
3   absent    2       5      1
4   absent    1       4     15
5   absent    1       2     16
6   absent   61       2     17
7   absent   37       3     16
8   absent  113       2     16
9  present   59       6     12


In [53]:
from sklearn.model_selection import train_test_split

In [54]:
#preparing features and target
X = Kyphosis_df.drop('Kyphosis',axis=1)
y = Kyphosis_df['Kyphosis']

In [55]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [56]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report

In [57]:
# selecting the right "parameters/hyperparameters of ML models can significantly affect their performacnce".
# let's begin exploring those parameters more for a DT model.
'''
DecisionTreeClassifier(
    criterion='gini',
    splitter='best',
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1,
    min_weight_fraction_leaf=0.0,
    max_features=None,
    random_state=None,
    max_leaf_nodes=None,
    min_impurity_decrease=0.0,
    min_impurity_split=None,
    class_weight=None,
    presort=False,
)
'''
# Let's consider these three parameters: criterion, splitter, min_samples_split
    # for parameters consider two cases 'gini' and 'entropy'
    # for splitter consider two cases 'best' and 'random'
    # for min_samples_split consider two cases 2 and 5
    # explain and dicuss your findings

"\nDecisionTreeClassifier(\n    criterion='gini',\n    splitter='best',\n    max_depth=None,\n    min_samples_split=2,\n    min_samples_leaf=1,\n    min_weight_fraction_leaf=0.0,\n    max_features=None,\n    random_state=None,\n    max_leaf_nodes=None,\n    min_impurity_decrease=0.0,\n    min_impurity_split=None,\n    class_weight=None,\n    presort=False,\n)\n"

In [66]:
def train_and_evaluate(criterion, splitter, min_samples_split):
    model = DecisionTreeClassifier(
        criterion=criterion,
        splitter=splitter,
        min_samples_split=min_samples_split,
    )
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    return accuracy

In [67]:
combinations = [
    ('gini', 'best', 2),
    ('gini', 'best', 5),
    ('gini', 'random', 2),
    ('gini', 'random', 5),
    ('entropy', 'best', 2),
    ('entropy', 'best', 5),
    ('entropy', 'random', 2),
    ('entropy', 'random', 5)
]

In [70]:
results={}
for criterion, splitter, min_samples_split in combinations:
    accuracy, report = train_and_evaluate(criterion, splitter, min_samples_split)
    key = f"criterion={criterion}, splitter={splitter}, min_samples_split={min_samples_split}"
    results[key] = accuracy

In [72]:
for key, accuracy in results.items():
    print(f"Parameters: {key}")
    print(f"Accuracy: {accuracy}")
    print("-" * 50)

Parameters: criterion=gini, splitter=best, min_samples_split=2
Accuracy: 0.84
--------------------------------------------------
Parameters: criterion=gini, splitter=best, min_samples_split=5
Accuracy: 0.8
--------------------------------------------------
Parameters: criterion=gini, splitter=random, min_samples_split=2
Accuracy: 0.84
--------------------------------------------------
Parameters: criterion=gini, splitter=random, min_samples_split=5
Accuracy: 0.68
--------------------------------------------------
Parameters: criterion=entropy, splitter=best, min_samples_split=2
Accuracy: 0.8
--------------------------------------------------
Parameters: criterion=entropy, splitter=best, min_samples_split=5
Accuracy: 0.76
--------------------------------------------------
Parameters: criterion=entropy, splitter=random, min_samples_split=2
Accuracy: 0.68
--------------------------------------------------
Parameters: criterion=entropy, splitter=random, min_samples_split=5
Accuracy: 0.76
-

In [74]:
best_combination = max(results, key=results.get)
print(f"\nBest combination: {best_combination}")
print(f"Best accuracy: {results[best_combination]}")


Best combination: criterion=gini, splitter=best, min_samples_split=2
Best accuracy: 0.84


Best Performing Combination:
The best performance was achieved with:

criterion: gini
splitter: best
min_samples_split: 2
This combination yielded an accuracy of 0.84 (84%).


Impact of Parameters:
a) Criterion (Gini vs Entropy):

Gini generally performed better than Entropy.
Gini's best accuracy: 0.84
Entropy's best accuracy: 0.80
This suggests that for this dataset, the Gini impurity measure was slightly more effective at creating meaningful splits.

b) Splitter (Best vs Random):

'Best' splitter generally outperformed 'Random'.
'Best' achieved the highest accuracy of 0.84 multiple times.
'Random' had more variable results, ranging from 0.68 to 0.84.
This indicates that for this dataset, systematically choosing the best split is more reliable than random splitting.

c) min_samples_split (2 vs 5):

A value of 2 generally performed better than 5.
With 2, we see the highest accuracies of 0.84.
With 5, the highest accuracy drops to 0.80.
This suggests that allowing the tree to create smaller splits (down to 2 samples) helps capture more nuanced patterns in this dataset.


Consistency:
The combination of Gini criterion with 'best' splitter seems most consistent, achieving 0.84 accuracy with both min_samples_split values.
Worst Performing Combinations:
The lowest accuracy (0.68) occurred twice:

Gini, Random, min_samples_split=5
Entropy, Random, min_samples_split=2
This reinforces that random splitting can lead to suboptimal performance, especially when combined with other constraining factors.


Overall Performance Range:
Accuracies ranged from 0.68 to 0.84, a difference of 16 percentage points. This significant range demonstrates the importance of parameter tuning for decision trees.
Robustness:
The 'best' splitter seems more robust to changes in other parameters, consistently achieving higher accuracies compared to 'random'.

Conclusions:

For this Kyphosis dataset, using the Gini criterion with the 'best' splitter and a min_samples_split of 2 yields the best performance.
The choice of splitter (best vs random) seems to have the most significant impact on performance.
Allowing for smaller splits (min_samples_split=2) generally improves performance, suggesting the presence of some fine-grained patterns in the data.
While Gini slightly outperforms Entropy, the difference is not drastic, indicating that both criteria are viable for this dataset.

Recommendations:

Use the best-performing combination (Gini, best, min_samples_split=2) for your final model.
If model simplicity is a concern, you could opt for min_samples_split=5 with Gini and 'best' splitter, as it still performs well (0.80 accuracy) and might result in a less complex tree.
Consider further tuning other parameters like max_depth or min_samples_leaf to potentially improve performance further.
If possible, gather more data to see if it helps improve the model's accuracy, especially given the relatively small difference between some parameter combinations.